File: csharp/src/DataStructures/Stack.cs
namespace DSA.DataStructures;

/// <summary>
/// Stack implementation using a linked list.
/// Time: O(1) for all operations
/// Space: O(n)
/// </summary>
public class Stack<T>
{
    private class Node
    {
        public T Value { get; }
        public Node? Next { get; set; }

        public Node(T value, Node? next = null)
        {
            Value = value;
            Next = next;
        }
    }

    private Node? _head;
    private int _size;

    public int Count => _size;
    public bool IsEmpty => _size == 0;

    public void Push(T value)
    {
        _head = new Node(value, _head);
        _size++;
    }

    public T? Pop()
    {
        if (_head == null)
            return default;

        var value = _head.Value;
        _head = _head.Next;
        _size--;
        return value;
    }

    public T? Peek()
    {
        return _head == null ? default : _head.Value;
    }

    public void Clear()
    {
        _head = null;
        _size = 0;
    }
}

File: php/src/DataStructures/Stack.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Stack implementation using a linked list.
 * Time: O(1) for all operations
 * Space: O(n)
 *
 * @template T
 */
class Stack
{
    private ?StackNode $head = null;
    private int $size = 0;

    public function push(mixed $value): void
    {
        $this->head = new StackNode($value, $this->head);
        $this->size++;
    }

    public function pop(): mixed
    {
        if ($this->head === null) {
            return null;
        }
        $value = $this->head->value;
        $this->head = $this->head->next;
        $this->size--;
        return $value;
    }

    public function peek(): mixed
    {
        return $this->head?->value;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function size(): int
    {
        return $this->size;
    }

    public function clear(): void
    {
        $this->head = null;
        $this->size = 0;
    }
}

/**
 * @internal
 */
class StackNode
{
    public function __construct(
        public mixed $value,
        public ?StackNode $next = null
    ) {}
}
